toolchain/check/impl_lookup.h

```diff
@@ -35,6 +38,60 @@ auto LookupImplWitness(Context& context, SemIR::LocId loc_id,
                        SemIR::ConstantId query_facet_type_const_id)
     -> SemIR::InstBlockIdOrError;
 
+// The result of EvalLookupSingleImplWitness(). It can be one of:
+// - No value. Lookup failed to find an impl declaration.
+// - A concrete value. Lookup found a concrete impl declaration that can be
+//   used definitively.
+// - A symbolic value. Lookup found an impl but it is not returned since lookup
+//   will need to be done again with a more specific query to look for
+//   specializations.
+class EvalImplLookupResult {
+ public:
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  EvalImplLookupResult(SemIR::InstId inst_id) : result_(inst_id) {}
+
+  static auto MakeNonFinalResult() -> EvalImplLookupResult {
+    return EvalImplLookupResult(FoundNonFinalImpl());
+  }
+
+  // True if a concrete impl witness was found or a non-final impl. In the
+  // latter case the InstId of the impl's witness is not returned, only the fact
+  // that it exists.
+  auto has_value() const -> bool {
+    return std::holds_alternative<FoundNonFinalImpl>(result_) ||
+           std::get<SemIR::InstId>(result_).has_value();
+  }
+
+  // True if there is a concrete witness in the result. If false, and
+  // `has_value()` is true, it means a non-final impl was found and a further
+  // more specific query will need to be done.
+  auto has_concrete_value() const -> bool {
```

Comment:
Maybe rename to match `GetConcreteWitness`?

```suggestion
  auto has_concrete_witness() const -> bool {
```